fdbserver/CommitProxyServer.actor.cpp

```diff
@@ -1063,6 +1071,88 @@ void assertResolutionStateMutationsSizeConsistent(const std::vector<ResolveTrans
 	}
 }
 
+// Return true if a single-key mutation is associated with a valid tenant id or a system key
+bool validTenantAccess(MutationRef m, std::map<int64_t, TenantName> const& tenantMap, Optional<int64_t>& tenantId) {
+	if (isSystemKey(m.param1))
+		return true;
+
+	if (isSingleKeyMutation((MutationRef::Type)m.type)) {
+		tenantId = TenantAPI::extractTenantIdFromMutation(m);
+		return tenantMap.count(tenantId.get()) > 0;
+	}
+	return true;
+}
+
+inline bool tenantMapChanging(MutationRef const& mutation) {
+	const KeyRangeRef tenantMapRange = TenantMetadata::tenantMap().subspace;
+	if (mutation.type == MutationRef::SetValue && mutation.param1.startsWith(tenantMapRange.begin)) {
+		return true;
+	} else if (mutation.type == MutationRef::ClearRange &&
+	           tenantMapRange.intersects(KeyRangeRef(mutation.param1, mutation.param2))) {
+		return true;
+	}
+	return false;
+}
+
+// Return success and properly split clear range mutations if all tenant check pass. Otherwise, return corresponding
+// error
+Error validateAndProcessTenantAccess(VectorRef<MutationRef> mutations,
+                                     ProxyCommitData* const pProxyCommitData,
+                                     Optional<UID> debugId = Optional<UID>(),
+                                     const char* context = "") {
+	bool changeTenant = false;
+	bool writeNormalKey = false;
+
+	for (auto& mutation : mutations) {
+		Optional<int64_t> tenantId;
+		bool validAccess = true;
+		if (tenantMapChanging(mutation)) {
+			changeTenant = true;
+		} else if (mutation.type == MutationRef::ClearRange) {
+			auto beginTenantId = TenantAPI::extractTenantIdFromKeyRef(mutation.param1);
+			auto endTenantId = TenantAPI::extractTenantIdFromKeyRef(mutation.param2);
+			CODE_PROBE(beginTenantId != endTenantId, "Clear Range raw access or cross multiple tenants");
+			// TODO: splitting clear range
+		} else if (!isSystemKey(mutation.param1)) {
+			validAccess = validTenantAccess(mutation, pProxyCommitData->tenantMap, tenantId);
```

Comment:
`isSystemKey` is also being checked by `validTenantAccess` so we can remove one of these checks.

Reply:
This is for `writeNormalKey` is correctly set.

Reply:
What I mean is that in line 1116, we are checking `isSystemKey(...)`. Then the first thing we do inside `validTenantAccess` is check `isSystemKey(...)`. We only need to do that once here

Reply:
To prevent incorrect use of `validTenantAccess`, we still need to do the `ASSERTION(!isSystemKey)` at the beginning. I doubt we can save a call here